File: 0046_permute.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        temp = []
        self.DFS(nums, result, temp)
        return result
    
    def DFS(self, nums, result, temp):
        if not nums:
            result.append(temp)
            return
        for i in range(len(nums)):
            self.DFS(nums[:i] + nums[i+1:], result, temp + [nums[i]])

# Alternative solution

class Solution1:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        self.backtrack(nums, result, [])
        return result
    
    def backtrack(self, nums, result, temp):
        if len(nums) == 0:
            result.append(temp.copy())
            return
        else:
            for i, n in enumerate(nums):
                temp.append(n)
                self.backtrack(nums[i+1:] + nums[:i], result, temp)
                temp.pop()
                
# Alternative solution

class Solution2:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        self.backtrack(nums, result, [])
        return result
    
    def backtrack(self, nums, result, temp):
        if len(nums) == 0:
            result.append(temp)
            return
        else:
            for i, n in enumerate(nums):
                self.backtrack(nums[i+1:] + nums[:i], result, temp + [n])

# Alternative solution

class Solution3:
    def permute(self, nums: List[int]) -> List[List[int]]:
        output = []
        element = []
        self.dfs(nums, output, element)
        return output
        
    def dfs(self, nums, output, element):
        if not nums:
            output.append(element)
            return
        for i in range(len(nums)):
            self.dfs(nums[i+1:] + nums[:i], output, element + [nums[i]])

File: 0001-0500/0046_permute.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> output = new ArrayList<>();
        backtrack(nums, output, new ArrayList<>(), new boolean[nums.length]);
        return output;
    }
    
    public void backtrack(int[] nums, List<List<Integer>> output, List<Integer> path, boolean[] used) {
        if (path.size() == nums.length) {
            output.add(new ArrayList<>(path));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (used[i] == true) {
                continue;
            }
            used[i] = true;
            path.add(nums[i]);
            backtrack(nums, output, path, used);
            path.remove(path.size() - 1);
            used[i] = false;
        }
    }
}
